src/sys/signal.rs

```diff
@@ -831,6 +831,127 @@ mod sigevent {
     }
 }
 
+/// A wrapper for `stack_t`, used to represent alternate stacks for `sigaltstack`.
+///
+/// See [`sigaltstack`](http://pubs.opengroup.org/onlinepubs/9699919799/functions/sigaltstack.html)
+/// details on the requirements for an alternate signal stack.
+#[derive(Copy, Clone)]
+#[allow(missing_debug_implementations)]
+pub struct SigStack {
+    stack: libc::stack_t,
+}
+
+impl SigStack {
+    /// Define an alternate signal stack for use with `sigaction` using a raw pointer.
+    ///
+    /// The memory pointed to by `sp` must remain valid for as long as any signal handlers are
+    /// registered to use this stack.
+    ///
+    /// ```
+    /// # use nix::libc;
+    /// # use nix::sys::mman::{mmap, munmap, MapFlags, ProtFlags};
+    /// # use nix::sys::signal::{sigaltstack, SigStack, SigStackFlags};
+    /// # use std::ptr;
+    /// let stack_mem = unsafe {
+    ///     mmap(
+    ///         ptr::null_mut(),
+    ///         libc::SIGSTKSZ,
+    ///         ProtFlags::PROT_READ | ProtFlags::PROT_WRITE,
+    ///         MapFlags::MAP_ANON | MapFlags::MAP_PRIVATE,
+    ///         -1,
+    ///         0,
+    ///     )
+    ///     .unwrap()
+    /// };
+    /// let sigstack = SigStack::from_raw(stack_mem, SigStackFlags::empty(), libc::SIGSTKSZ);
+    /// unsafe {
+    ///     let old_sigstack = sigaltstack(&sigstack).unwrap();
+    ///     sigaltstack(&old_sigstack).unwrap();
+    ///     // unmap memory only after the stack is unregistered
+    ///     munmap(stack_mem, libc::SIGSTKSZ);
+    /// }
+    /// ```
+    pub fn from_raw(sp: *mut libc::c_void, flags: SigStackFlags, size: libc::size_t) -> SigStack {
```

Comment:
I think that the only way to create the SigStack should be from a slice.  It's a cleaner interface than using the raw pointer, and it's easy for the user to construct his own slice if he must.

Reply:
Constructing a slice as an intermediate step introduces an unnecessary use of `unsafe`, so I'd much rather keep `from_raw`.